### PR TITLE
CVSL-1906 Pull approval case list building logic into service

### PR DESCRIPTION
**Pull request by andrewrlee**

### CODE REVIEW

File: server/services/approverCaseloadService.ts [hunk 1]
```diff
@@ -1,56 +1,57 @@
...
 
-  async getApprovalNeeded(user: User, prisonCaseload: string[]): Promise<ManagedCaseForApproval[]> {
+  async getApprovalNeeded(user: User, prisonCaseload: string[], searchString: string): Promise<ApprovalCase[]> {
     const licences: LicenceSummaryApproverView[] = await this.licenceApiClient.getLicencesForApproval(
       filterCentralCaseload(prisonCaseload),
       user
     )
     const caseLoad: ManagedCaseForApproval[] = await this.mapLicencesToOffenders(licences)
-    return this.mapResponsibleComsToCases(caseLoad)
+    return this.applySearch(searchString, await this.mapResponsibleComsToCases(caseLoad))
```
Comment (andrewrlee) [inline server/services/approverCaseloadService.ts:45]: search filtering is now done as an after step

File: server/services/approverCaseloadService.ts [hunk 2]
```diff
@@ -1,56 +1,57 @@
...
+  async getRecentlyApproved(user: User, prisonCaseload: string[], searchString: string): Promise<ApprovalCase[]> {
     const licences: LicenceSummaryApproverView[] = await this.licenceApiClient.getLicencesRecentlyApproved(
       filterCentralCaseload(prisonCaseload),
       user
     )
     const caseLoad: ManagedCaseForApproval[] = await this.mapLicencesToOffenders(licences)
-    return this.mapResponsibleComsToCases(caseLoad)
-  }
-
-  private pairDeliusRecordsWithNomis = async (
```
Comment (andrewrlee) [inline server/services/approverCaseloadService.ts:35]: just moved this method unchanged down a bit

File: server/services/approverCaseloadService.ts [hunk 3]
```diff
@@ -97,38 +119,74 @@ export default class ApproverCaseloadService {
 
     const coms = await this.communityService.getStaffDetailsByUsernameList(comUsernames)
 
-    return caseload.map(offender => {
```
Comment (andrewrlee) [inline server/services/approverCaseloadService.ts:100]: slight formatting change here - which is a bit confusing but code remains the same

File: server/services/approverCaseloadService.ts [hunk 4]
```diff
@@ -97,38 +119,74 @@ export default class ApproverCaseloadService {
...
-            name: `${responsibleCom.staff.forenames} ${responsibleCom.staff.surname}`.trim(),
+            staffCode: offender.deliusRecord.staff.code,
+            name: `${offender.deliusRecord.staff.forenames} ${offender.deliusRecord.staff.surname}`.trim(),
           },
         }
-      }
-
-      if (!offender.deliusRecord.staff || offender.deliusRecord.staff.unallocated) {
+      })
+      .map((c: ManagedCaseForApproval) => {
```
Comment (andrewrlee) [inline server/services/approverCaseloadService.ts:156]: performing the same mapping that occurred in the frontend

File: server/services/approverCaseloadService.ts [hunk 5]
```diff
@@ -97,38 +119,74 @@ export default class ApproverCaseloadService {
...
+          releaseDate: releaseDate ? format(releaseDate, 'dd MMM yyyy') : 'not found',
+          sortDate: releaseDate,
+          urgentApproval,
+          approvedBy: licence.approvedBy,
+          approvedOn: approvedDate,
+          isDueForEarlyRelease: licence.isDueForEarlyRelease,
         }
-      }
+      })
+      .sort((a, b) => (a.sortDate?.getTime() || 0) - (b.sortDate?.getTime() || 0))
```
Comment (andrewrlee) [inline server/services/approverCaseloadService.ts:178]: and the sorting too

File: server/routes/approvingLicences/handlers/approvalCases.ts [hunk 6]
```diff
@@ -20,50 +17,15 @@ export default class ApprovalCaseRoutes {
     const allPrisons = await this.prisonerService.getPrisons()
     const activeCaseload = allPrisons.filter(p => p.agencyId === user.activeCaseload)
     const prisonCaseloadToDisplay = caseloadsSelected.length ? caseloadsSelected : [activeCaseload[0].agencyId]
+    const searchTerm = search?.toLowerCase().trim()
     const cases = approvalNeededView
-      ? await this.approverCaseloadService.getApprovalNeeded(user, prisonCaseloadToDisplay)
```
Comment (andrewrlee) [inline server/routes/approvingLicences/handlers/approvalCases.ts:24]: all this logic is now in the service

File: server/routes/approvingLicences/handlers/approvalCases.test.ts [hunk 7]
```diff
@@ -135,394 +93,57 @@ describe('Route Handlers - Approval - case list', () => {
...
-        search: undefined,
-        approvalNeededView: true,
-      })
-    })
-
-    it('should prioritise licence release date if present', async () => {
-      const caseLoadData = [
-        {
-          licences: [
+    describe('GET approval needed', () => {
```
Comment (andrewrlee) [inline server/routes/approvingLicences/handlers/approvalCases.test.ts:96]: the tests here are much simpler now -
```
get licences to approve
  render called correctly
  search filter provided
get recently approved licences
  render called correctly
  search filter provide
```

File: server/services/approverCaseloadService.test.ts [hunk 8]
```diff
@@ -25,164 +31,248 @@ describe('Approval Caseload Service', () => {
...
-        ],
-        probationPractitioner: {
-          staffCode: 'X1234',
-          name: 'Joe Bloggs',
+      const result = await serviceUnderTest.getRecentlyApproved(user, [], undefined)
+
+      expect(result).toStrictEqual([expectedResult])
+    })
+
+    it('handles missing release dates', async () => {
```
Comment (andrewrlee) [inline server/services/approverCaseloadService.test.ts:147]: extra tests for searching / sorting etc..

File: server/services/approverCaseloadService.ts (hunk 4 above)
Comment (andrewrlee) [reply server/services/approverCaseloadService.ts:156]: (the caseload building code here is now incredibly complex but will simplify in a follow up)

**Review by nishanthmahasamudram — APPROVED** [review]
LGTM

**Pull request merged**